assets/js/atomic/blocks/product-elements/button/frontend.tsx

```diff
@@ -0,0 +1,151 @@
+	woocommerce: {
+		addToCartText: ( {
+			context,
+			state,
+		}: {
+			context: Context;
+			state: State;
+		} ) => {
+			if ( context.woocommerce.numberOfItems === 0 ) {
+				return context.woocommerce.addToCartText;
+			}
+
+			if ( ! state.woocommerce.cart ) {
+				return state.woocommerce.inTheCartText.replace(
+					'###',
+					context.woocommerce.numberOfItems.toString()
+				);
+			}
+
+			const cartState = state.woocommerce.cart;
+			const product = getProductById(
+				cartState,
+				context.woocommerce.productId
+			);
+
+			if ( ! product ) {
+				return context.woocommerce.addToCartText;
+			}
+
+			return state.woocommerce.inTheCartText.replace(
+				'###',
+				product?.quantity?.toString()
+			);
+		},
+		isThereMoreThanOneItem: ( {
+			context,
+			state,
+		}: {
+			context: Context;
+			state: State;
+		} ) => {
+			if ( ! state.woocommerce?.cart ) {
+				return context.woocommerce.numberOfItems > 0;
+			}
+			const product = getProductById(
+				state.woocommerce.cart,
+				context.woocommerce.productId
+			);
+
+			return product !== undefined && product.quantity > 0;
+		},
+		isAdded: ( {
+			context,
+			selectors,
+			state,
+		}: {
+			context: Context;
+			// For now, let's just use any.
+			// eslint-disable-next-line @typescript-eslint/no-explicit-any
+			selectors: any;
+			state: State;
+		} ) => {
+			return selectors.woocommerce.isThereMoreThanOneItem( {
+				context,
+				state,
+			} );
+		},
+	},
+};
+
+// @ts-expect-error: Store function isn't typed.
+store( {
+	selectors: productButtonSelectors,
+	actions: {
+		woocommerce: {
+			addToCart: async ( {
+				context,
+				ref,
+			}: {
+				context: Context;
+				state: State;
+				ref: HTMLElement;
+			} ) => {
+				if ( ! ref.classList.contains( 'ajax_add_to_cart' ) ) {
+					return;
+				}
+
+				context.woocommerce.isLoading = true;
+				const body = document.body;
+
+				// Allow 3rd parties to validate and quit early. https://github.com/woocommerce/woocommerce/blob/154dd236499d8a440edf3cde712511b56baa8e45/plugins/woocommerce/client/legacy/js/frontend/add-to-cart.js/#L74-L77
+				const event = new CustomEvent(
+					'should_send_ajax_request.adding_to_cart',
+					{ detail: [ ref ], cancelable: true }
+				);
+				const shouldSendRequest = body.dispatchEvent( event );
+
+				if ( shouldSendRequest === false ) {
+					const ajaxNotSentEvent = new CustomEvent(
+						'ajax_request_not_sent.adding_to_cart',
+						{ detail: [ false, false, ref ] }
+					);
+					body.dispatchEvent( ajaxNotSentEvent );
+					return true;
+				}
+
+				try {
+					await dispatch( storeKey ).addItemToCart(
+						context.woocommerce.productId,
+						1
+					);
+					context.woocommerce.numberOfItems++;
+					context.woocommerce.isLoading = false;
+				} catch ( error ) {
+					// we don't care about errors blocking execution, but will console.error for troubleshooting.
+					// eslint-disable-next-line no-console
+					console.error( error );
```

Comment:
Nice catch! Given that we are using the Store API, we can't rely on the PHP to generate the notice and show it on the product page. What do you think about the approach that I introduced with [54cb0ff](https://github.com/woocommerce/woocommerce-blocks/pull/10006/commits/54cb0ffe29356f1949fefc8d1e0b3cae256b193c)?

https://github.com/woocommerce/woocommerce-blocks/assets/4463174/6e79f635-9354-4575-8a29-49657d50eaf8

Reply:
Looks perfect! The only issue I found is that the notice doesn't show up if the Store Notices block is not on the page. I think that would be an issue when adding the Products block to a post or page, for example, as the Store Notices block can't be added in that context.

Reply:
Yup, makes sense to me. We also have the [`StoreNoticesContainer` component](https://github.com/woocommerce/woocommerce-blocks/blob/55cc42e761a5c509bb2c5fae95cd241c497c213d/packages/checkout/components/store-notices-container/index.tsx#L29), but I don't know if it would be too much for this use-case.

Reply:
I improved the logic with 72ff24763471494c97837bdb45caa8e10b2854d9

Reply:
Is this something that the upcoming auto-insert mechanism could handle?

- https://github.com/WordPress/gutenberg/pull/51449

If not, could it use the Interactivity API instead of a React component?

Reply:
Oh, I thought it was a client-side component. Sorry.